src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h

```diff
@@ -677,8 +692,162 @@ namespace OpenMS
       return EXECUTION_OK;
     }
 
-protected:
-    struct PeptideProteinMatchInformation
+ protected:
+     using DecoyStringToAffixCount = std::map<std::string, std::pair<int, int>>;
+     using CaseInsensitiveToCaseSensitiveDecoy = std::map<std::string, std::string>;
+
+     template<typename T>
+     bool findDecoyString_(FASTAContainer<T>& proteins,
+                                std::string& decoy_string,
+                                bool& is_prefix)
+     {
+       // Common decoy strings in fasta files
+       // Note: decoy prefixes/suffices must be provided in lower case and must only contain alphanumeric characters
+        DecoyStringToAffixCount decoy_count = {{"decoy", {0,0}}, {"reverse", {0,0}},
+                                                          {"rev", {0,0}}, {"dev", {0,0}},
+                                                          {"iddecoy", {0,0}}, {"xxx", {0,0}},
+                                                          {"shuffled", {0,0}}, {"shuffle", {0,0}},
+                                                          {"pseudo", {0,0}}, {"random", {0, 0}}};
+       // Map case insensitive strings back to original case (as used in fasta)
+       CaseInsensitiveToCaseSensitiveDecoy decoy_case_sensitive;
+
+       // Setup regex vector, sort descending
+       std::vector<std::string> prefix_regex;
+       for (const auto& pair : decoy_count)
+       {
+         prefix_regex.push_back(pair.first);
+       }
+       std::sort(prefix_regex.rbegin(), prefix_regex.rend());
+
+       // Setup regex string -> create prefix- and suffix pattern to match to
+       std::string regexstr_prefix = "^(";
```

Comment:
how about `std::string regexstr_prefix = "^(" + ListUtils::concatenate<String>(prefix_regex, "|") + ")";` :)

Reply:
That's pretty!